### ISSUE TITLE
(feature request summary)

### ISSUE THREAD

**Issue by Kyllingene** [vlang/v#16957]
### Describe the feature

In Rust, you can do things like this:

```rust
fn do_something(mut i: i32) {
	// ...
	i += 1;
	// ...
}
```

In this code, the variable `i` is only changed for the function `do_something`, since `i` is not a mutable reference, just a mutable variable.

### Use Case

 This avoids messy code like this:

```v
fn do_something(a int, b int, c int, d int) {
	mut i := a;
	mut j := b;
	mut k := c;
	mut l := d;
}
```

This is an extreme example, and derived from some code that should probably use a struct. However, the principle is the same, and I use mutable function arguments frequently in Rust.

### Proposed Solution

The implementation is up for debate. Simply changing how the `mut` keyword works when applied to a non-`arrays, interfaces, maps, pointers, structs or their aliases` function argument could be confusing, since it would then alternate between modifying data for the caller, and modifying it only for itself; however, I don't think its much more confusing than the reference semantics already are (`mut` implies a reference, without explicitly making one, while without the `mut` it could do either).

The alternatives are making a new keyword for primitive arguments, or changing the current `mut` syntax to require an explicit reference. Neither of these seem like good options to me.

I looked at `v/vlib/v/parser/fn.v`, and it looked like atomics were already mutable, but 

### Other Information

To me, this seems like a symptom of slightly confusing syntax that has become settled. I like the V motto of `don't worry about it`, but in this case it leads to potential ambiguities and lost opportunities.

A possible alternative would be allowing redefinitions, but that doesn't look like it's going to happen. That would also be missing what (to me) is the core issue, that being the `mut` syntax.

Another solution would be allowing you to change primitives by default, without requiring the `mut` keyword at all. The issue then is that it breaks from the tradition of having to make mutable variables explicit. The upside is that it seems relatively easy to implement.

### Acknowledgements

- [X] I may be able to implement this feature request
- [X] This feature might incur a breaking change

### Version used

V 0.3.2 3f8aa77.d9339b0

### Environment details (OS name and version, etc.)

```
OS: linux, "Manjaro Linux"
Processor: 12 cpus, 64bit, little endian, AMD Ryzen 5 2600 Six-Core Processor
CC version: cc (GCC) 12.2.0

getwd: /home/ME
vmodules: /home/ME/.vmodules
vroot: /home/ME/Apps/v
vexe: /home/ME/Apps/v/v
vexe mtime: 2023-01-11 23:37:42
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.2 3f8aa77.d9339b0

Git version: git version 2.39.0
Git vroot status: weekly.2023.02-6-gd9339b09 (10 commit(s) behind V master)
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3
```

**Comment by JalonSolov** [comment]
Changing primitives by default goes completely against V's ethos.  You are _supposed_ to consciously decide if something should be mutable or not, instead of just being able to change whatever you wish with no consequence (other than the to be expected debugging nightmares).

Immutability by default is a safety feature, whether it is convenient or not.

**Comment by ArtemkaKun** [comment]
I also disagree with that. Function arguments should be read-only or read/write-by-reference. This is a logical gap between "A data I need to operate on" and "A data I calculated based on data I received"

**Comment by spytheman** [comment]
> or changing the current mut syntax to require an explicit reference

I would have much preferred that, for clarity reasons, and so that the compiler and the programmer could always agree whether something is passed by reference, and the level of indirection, even though it would be a big breaking change.

Unfortunately it seems that it is unlikely to happen :-| .

**Comment by Larpon** [comment]
Could not `vfmt` handle a change like:
`fn (mut a Abc) def()` -> `fn (mut a &Abc) def()` since currently `mut` implies "reference"?
